bridges/snowbridge/docs/v2.md

```diff
@@ -0,0 +1,308 @@
+3. The same interface is used for both PNA and ERC20 tokens. Internally we will still look up whether the token is registered as a PNA or ERC20 for the purpose of minting/locking burning/unlocking logic. The asset transfer type chosen by the UX layer will inform the XCM that is built from the message on BH.
+
+```solidity
+enum Kind {
+    Index,
+    Address32,
+    Address20,
+    XCMPayload,
+}
+
+struct Beneficiary {
+    Kind kind;
+    bytes data;
+}
+
+enum AssetTransferType {
+	LocalReserve, DestinationReserve, Teleport
+}
+
+struct Token {
+	AssetTransferType type;
+	address token;
+	uint128 amount;
+}
+
+function send(
+  ParaID destinationChain,
+  Beneficiary calldata beneficiary,
+	Token[] tokens,
+  uint128 reward) external payable;
+```
+
+Message enqueued $m_0$:
+
+```solidity
+send(
+	3022, // KILT Para Id
+	Address32(0x0000....),
+	[
+		(DestinationReserve, KLT, 100)
+	],
+	10, // WETH
+	)
+```
+
+```solidity
+send { value: 3 }( // Send 3 Eth for fees and reward
+	3022, // KILT Para Id
+	XCMPayload(
+		DepositAsset (KLT, 100) dest=Bob
+		Transact Bob.hello()
+	),
+	[
+		(DestinationReserve, KLT, 100)
+	],
+	1, // 1 ETH of 3 needs to be for the reward, the rest is for fees
+	)
+```
+
+The User Agent/UX layer will need to estimate the fee required to be passed into the `send` method. This may be an issue as we cannot Dry-Run something on Polkadot that has not even been submitted on Ethereum yet. We may need to make RPC API to DryRun and get back the xcm that would be submitted to asset hub.
+
+## Step 2: Relayer relays message to Bridge Hub
+
+Exchange rate is eliminated. Users pay remote delivery costs in ETH, and this amount is sent with the message as WETH. The delivery fee can be claimed by the relayer on BH.
+
+The user agent applies a similar dry-running process as with [Step 2: User agent estimates fees](https://www.notion.so/Step-2-User-agent-estimates-fees-113296aaabef8159bcd0e6dd2e64c3d0?pvs=21).
+
+The message is converted from $m_0$ to $x_0$ during message submission on BH. Dry-running submission will return $x_0$ to the relayer so that it can verify it is profitable.
+
+## Step 3: AH receives $x_0$ from BH
+
+Submitting the message $m_0$ will cause the following XCM, $x_0$, to be built on BH and dispatched to AH.
+
+```
+WithdrawAsset (KLT, 100)
+ReserveAssetDeposited(WETH, 2)
+PayFees (WETH, 1)
+SetAssetClaimer Bob            // derived from beneficiary
+AliasOrigin Ethereum/Alice     // derived from msg.sender
+InitiateAssetsTransfer asset=(KLT, 100) remoteFee=(WETH, 1) dest=KLT
+	DepositAsset (KLT, 100) dest=Bob
+```
+
+```
+WithdrawAsset (KLT, 100)
+ReserveAssetDeposited(WETH, 2)
+PayFees (WETH, 1)
+SetAssetClaimer Ethereum/Alice   // derived from beneficiary
```

Comment:
Yes, this would be derived by the msg.sender on Ethereum. If the user interacts with the bridge indirectly through a smart contract than it will be that contracts address. If the user calls the bridge directly, through a dapp, then it will be the user address.